vpr/src/route/connection_router.cpp

```diff
@@ -649,27 +755,47 @@ void ConnectionRouter<Heap>::evaluate_timing_driven_node_costs(t_heap* to,
 
     float total_cost = 0.;
     const t_conn_delay_budget* delay_budget = cost_params.delay_budget;
-    if (delay_budget) {
+
+    if (run_rcv && to->path_data != nullptr) {
+        to->path_data->backward_delay += cost_params.criticality * Tdel;
+        to->path_data->backward_cong += (1. - cost_params.criticality) * get_rr_cong_cost(to_node, cost_params.pres_fac);
+
+        float expected_delay; // = router_lookahead_.get_expected_delay(to_node, target_node, cost_params, to->R_upstream);
+        float expected_cong;  // = router_lookahead_.get_expected_cong(to_node, target_node, cost_params);
+
+        std::tie(expected_delay, expected_cong) = router_lookahead_.get_expected_delay_and_cong(to_node, target_node, cost_params, to->R_upstream);
+
+        float expected_total_delay_cost; // = new_costs.backward_cost + cost_params.astar_fac * expected_cost;
+        float expected_total_cong_cost;
+
+        float expected_total_cong = cost_params.astar_fac * expected_cong + to->path_data->backward_cong;
+        float expected_total_delay = cost_params.astar_fac * expected_delay + to->path_data->backward_delay;
         //If budgets specified calculate cost as described by RCV paper:
         //    R. Fung, V. Betz and W. Chow, "Slack Allocation and Routing to Improve FPGA Timing While
         //     Repairing Short-Path Violations," in IEEE Transactions on Computer-Aided Design of
         //     Integrated Circuits and Systems, vol. 27, no. 4, pp. 686-697, April 2008.
 
         //TODO: Since these targets are delays, shouldn't we be using Tdel instead of new_costs.total_cost on RHS?
-        total_cost += (delay_budget->short_path_criticality + cost_params.criticality) * std::max(0.f, delay_budget->target_delay - total_cost);
-        total_cost += std::pow(std::max(0.f, total_cost - delay_budget->max_delay), 2) / 100e-12;
-        total_cost += std::pow(std::max(0.f, delay_budget->min_delay - total_cost), 2) / 100e-12;
-    }
-
-    //Update total cost
-    float expected_cost = router_lookahead_.get_expected_cost(to_node, target_node, cost_params, to->R_upstream);
-    VTR_LOGV_DEBUG(router_debug_ && !std::isfinite(expected_cost),
-                   "        Lookahead from %s (%s) to %s (%s) is non-finite, expected_cost = %f, to->R_upstream = %f\n",
-                   rr_node_arch_name(to_node).c_str(), describe_rr_node(to_node).c_str(),
-                   rr_node_arch_name(target_node).c_str(), describe_rr_node(target_node).c_str(),
-                   expected_cost, to->R_upstream);
-    total_cost += to->backward_path_cost + cost_params.astar_fac * expected_cost;
 
+        expected_total_delay_cost = expected_total_delay;
+        expected_total_delay_cost += (delay_budget->short_path_criticality + cost_params.criticality) * std::max(0.f, delay_budget->target_delay - expected_total_delay);
```

Comment:
I believe you should test if std::max(0.f ...) should becoming an if statement like:
```
float diff = delay_budget->target_delay - expected_total_delay;
if(diff  > 0) {
   expected_total_delay_cost += (delay_budget->short_path_criticality + cost_params.criticality) * diff;
} 
```